### ISSUE TITLE
"adapt" should use a `holdSubject`

### ISSUE THREAD

**Issue by ntilwalli** [cyclejs/cyclejs#373]
While playing around with `@cycle/storage` I noticed I was missing the initial emission which happens upon calling `getItem` and tracked it down to the fact that those emissions happen synchronously during inital wiring of the app before any listeners have been added.  Calling `runSA.remember(runSA.adapt(...)) doesn't fix the issue because the adapt call causes the initial `addListener` call which emits before the `thru(hold) has been wired in, so the initial emission drops.  While researching this I discovered follow on issues with `most-subject` (https://github.com/mostjs-community/most-subject/issues/22 and https://github.com/mostjs-community/most-subject/issues/23), but in theory switching `adapt` to use a `holdSubject` would  fix this issue (and is what I've currently hacked in).

**Comment by ntilwalli** [comment]
@staltz The initial emission on a stream is not being sent to a requesting listener.  Why is that not a bug?

**Comment by staltz** [comment]
I marked it feature suggestion temporarily until I have time to look at this carefully and decide if it's a bug. I remember reasoning about this together with Tylor and we got to the conclusion we never need holdSubject, or at least we couldn't come up with an example where we needed it, and therefore we removed it because we don't want to have code which "we think we may need it but we don't know exactly why".

**Comment by ntilwalli** [comment]
Got it.  Thanks for the response.  This seems like an example of a situation where it's required.  `addListener/subscribe/forEach...` is called on a stream which synchronously emits it's first element using `startWith` or whatever and is sent through a basic subject (no memory/replay/hold...) in via the stream adapter's `adapt` that has no listeners yet.  The `remember` call happens afterwards so that initial emission gets dropped.

Personally, instead of changing `adapt` I actually think the stream adapter's `remember` method should use a createWithMemory/ReplaySubject/holdSubject... internally.  Then driver authors can choose to call `remember` instead of `adapt` if their stream emits synchronously on initial subscription.

**Comment by whitecolor** [comment]
> Personally, instead of changing adapt

> remember method should use a createWithMemory/ReplaySubject/holdSubject... internally. 

Adapt is abstract. You can pass to adapt what ever `originStreamSubscribe` you wan't. It would be better to see your code (is it a driver?) that doesn't work.

**Comment by ntilwalli** [comment]
Here's an example:
```
import most from 'most'
import Cycle from '@cycle/most-run'
import storageDriver from '@cycle/storage'

function main(sources) {
  const {Storage} = sources
  Storage.local.getItem(`nonExistentItem`).forEach(x => console.log(x))
  return {
    Storage: most.never()
  }
}

Cycle.run(main, {
  Storage: storageDriver()
})
```

The `getItem` stream emits "{}" initially since the requested item doesn't exist and it is not printed since it's currently being dropped during `adapt`.

**Comment by staltz** [comment]
@ntilwalli I'm trying to understand this bug but getting confused. I think you initially reported the *solution* (because you found a hack), but not exactly the problem. For instance you mentioned the `adapt()` call, but we use that in many places, I don't know which usage of it are you referring to.

- What is your application code like?
- What was the expected behavior?
- What was the actual behavior?

**Comment by ntilwalli** [comment]
Application code: See above posted code (I just fixed some typos BTW... `storageDriver() -> storageDriver`, `{}` -> `null`)

Expected behavior: prints `null` once
Actual behavior: prints nothing

This is an odd bug in that even if the stream adapter is changed in the way I'm requesting it still would not print `null` because of another bug in `most-adapter` which I've requested to have fixed here: https://github.com/mostjs-community/most-subject/issues/22

You could build that package from my repository which has the proposed fix (https://github.com/ntilwalli/most-subject) but that's annoying.

To reiterate the problem: the initial emission from the `getItem` stream is being dropped

Current behavior:
1) User calls `Storage.local.getItem('someItem')`
2) `getItem` internally generates a stream with a `startWith()` which we'll call X, which is then passed through the stream adapter's `adapt` call. 
3) `adapt` takes X and internally creates a regular `subject` which we'll call Y
4) Internally `adapt` has the side effect of triggering an `addListener` call on X
5) `addlistener` triggers the `startWith` to emit immediately (let's call the sent message MSG)
6) `next` is called on Y with MSG, but it has no listeners, so MSG is **dropped**.
7) Y is returned to main as the stream
8) Cycle.run wires up the sinks and sources including Y as normal and the application code never sees MSG

This would not be fixed if `setItem` was changed to return something like `remember(adapt(...))` since the adapt is where the side effect takes place which is causing the issue so the stream created by the `remember` call, meant to capture the emission would not have been created yet.

A holdSubject would solve this problem since it would buffer the emission even though there are no listeners during the `adapt` call.

6)
4)

User

**Comment by ntilwalli** [comment]
BTW, to test this you need the latest version of `@cycle/storage`.  The visible version on npm is out of date, you can install the diversity version using `npm install --save @cycle/storage@3.0.0-rc3`

I've created an issue: https://github.com/cyclejs/storage/issues/19

**Comment by whitecolor** [comment]
> This would not be fixed if setItem was changed to return something like `remember(adapt(...))`

So this should be done, no adapter change required.

**Comment by ntilwalli** [comment]
@whitecolor what I wrote was calling `remember(adapt(...))` would **not** fix the issue because adapt would be called, the side effecting `addListener` call would happen, the item would be emitted, dropped _and then_ the call to `remember` would take place, which would be too late to buffer the initial emission.

**Comment by whitecolor** [comment]
Where this `adapt` is used currenlty? In what driver?

**Comment by ntilwalli** [comment]
https://github.com/cyclejs/storage/blob/diversity/src/util.js#L45

**Comment by whitecolor** [comment]
Probably try:
```
      return runSA.adapt(
        XStreamAdapter.remember(storageGetItem(key, request$, type)),
        XStreamAdapter.streamSubscribe        
      )
```

**Comment by ntilwalli** [comment]
That won't fix it.  Even if the internal stream is remembered, when the `addListener` method is invoked inside `adapt` stream will emit into a basic subject which will not remember the value since there are no listeners.  The stream that's exposed to the world is not the internal stream, it's the subject returned by `adapt`.  The subject itself requires memory, not the stream that it is subscribing to, hence the need for a holdSubject.

**Comment by whitecolor** [comment]
May try this, should subscribe to getItem stream only after subject is subscribed:
```js
 return streamAdapter.adapt({}, (_, observer) => 
   XStreamAdapter.streamSubscribe(storageGetItem(key, request$, type), observer)
  )  
```

**Comment by ntilwalli** [comment]
That might work but it doesn't seem elegant.  @staltz is the above proposal the preferred way to manage `startWith` emissions in drivers?

**Comment by whitecolor** [comment]
> That might work but it doesn't seem elegant.

Well right, this is the issue actually specific of most-adapter (runSA), others shouldn't have it. Probably you are right in this case that HoldSubject is needed to get latest value - hot streams, they are so hot.

**Comment by ntilwalli** [comment]
@whitecolor the more I think about it, the more I like your proposal.  What I don't like is driver writers needing to code the deferral.  Why doesn't `adapt` do that by default now?  It should.  The stream returned by `adapt` should only call the internal `addListener` upon seeing it's first `subscribe` call, which would make it act like any standard source.

**Comment by ntilwalli** [comment]
@whitecolor why wouldn't other stream libraries have this issue?  It seems to me any time you're crossing a library boundary using a stream adapter this kind of issue could come up.

**Comment by whitecolor** [comment]
> why wouldn't other stream libraries have this issue?

Because in described case (in storage driver) with most-adapter you subscribe to original stream and send next to adapted stream before it is subscribed, and in other adapters you will do this only after adapted stream is subscribed/started. See the code of other adapters (rx, xstream).

**Comment by whitecolor** [comment]
And I think to bring most-adapter's `adapt` behaviour aligned with other adapters Hold Subject would be not enough: what you send **more then one** values synchronously? holdSubject if I get it right will "remember" only one. So there should be done something different inside the `adapt` method.

**Comment by ntilwalli** [comment]
Actually holdSubject takes a bufferSize parameter, but no other stream adapter offers that kind of buffer size flexibility so hard coding it to one in `most-adapter` would be fine.

**Comment by whitecolor** [comment]
I think there should some thing different it can be completed with subject I'm sure, so that would be the same as in other adapters as I wrote:

> you subscribe to original stream and send next to adapted stream before it is subscribed, and in other adapters you will do this only after adapted stream is subscribed/started

**Comment by ntilwalli** [comment]
I posted a PR.  I opted to make the subscription lazy instead of using a `holdSubject`.  Also dispose was not being called when the number of listeners went to zero, it was only called on completion, I fixed that.

**Comment by milankinen** [comment]
Nice notice! This is definitely a bug in `most-adapter`. 

The problem is not actually caused by the synchronized start of source stream but because it's subscription is [created eagerly](https://github.com/cyclejs/cyclejs/blob/master/most-adapter/src/index.ts#L19) in the adapter. The bug you @ntilwalli noticed is just a symptom and using `HoldSubject` won't fix it. There are other more critical problems because of this: if you create *any* stream by using driver API inside `main`, it immediately gets subscribed and never gets disposed, see logs from http://www.webpackbin.com/NJ-DoIytW

Making the subscription lazy with `@most/create` could solve the issue. Something like this (coded ad-hoc, may contain some errors):

```ts
import create from "@most/create"
...
  adapt<T>(originStream: any, originStreamSubscribe: StreamSubscribe): Stream<T> {
    return create((next, complete, error) =>  {
      return originStreamSubscribe(originStream, {next, complete, error})
    })
  }
```

**Comment by ntilwalli** [comment]
@milankinen did you look at the PR I posted to fix this issue? It addresses the problem you described directly.  It doesn't use @most/create but it works.

**Comment by whitecolor** [comment]
This should be a correct fix, didn't know that `create` is available via separate repo of `most`.

**Comment by whitecolor** [comment]
> EDIT: It seems PR does exactly this and contains no HoldSubject. Great job @ntilwalli!

are you talking about the PR with bunch of `setTimeout`s? I'm not sure that this a correct approach at all.

**Comment by ntilwalli** [comment]
@milankinen thanks!

If the code is changed to use `@most/create` that is fine but it doesn't change the fact that the disposable that is returned via `run` **needs** to be set asynchronously, for reasons discussed in the PR.  Either `setTimeout` needs to be used or an asap task needs to be added to the most scheduler. 

@whitecolor I responded to your concern about using `setTimeout` in the PR.  If you or someone know how to use the most scheduler to add an asap task, then please propose that change.  In lieu of that `setTimeout` is working as a near-term approach.

**Comment by milankinen** [comment]
>disposable that is returned via run needs to be set asynchronously, for reasons discussed in the PR

If I understood correctly, `@most/create` should do all that async stuff behalf of you so there is no need for additional setTimeout / asap task creation.

**Comment by ntilwalli** [comment]
When `adapt` is called it will call something like
```
const stream = create((add, end, error) => {
   const disposer = streamSubscribe(stream, {
     next: x => add(x),
     error: e => error(e),
     complete: () => end()
  })

  return disposer
})
```
In most, the disposer needs to be returned **before** the emissions begin.  Without a `setTimeout` or asap task to the scheduler, it's possible that the call to `streamSubscribe` will generate a synchronous emission immediately,  before the disposer has not been returned.  That is not allowed.  The disposer _must_ be returned before the stream is subscribed since sync emissions are possible when adding a listener to the adapted stream.  The explanation for this is given in the PR.

This issue arises both scenarios, 1) `@most/create` or 2) the current implemented custom source.  `streamSubscribe` must be called asynchronously so dispose can be guaranteed to be returned before the first emission.

**Comment by whitecolor** [comment]
First of all I want to notice that `setTimeout` should not be used here because it is not supposed to be used for such tasks, `setTimeout` is for relaxing and making things happen sparsely in time.  It should not be used for "scheduling" of sync tasks - this is always a code smell and a point of potential problems - it is really bad bad karma - it should be avoided and considered very carefully.

@ntilwalli 
> In lieu of that setTimeout is working as a near-term approach.

This doesn't mean that it should be in the adapter. You may want to include such delay in the driver if it really needs it.

This code:
```js
const stream = create((add, end, error) => {
   const disposer = streamSubscribe(stream, {
     next: x => add(x),
     error: e => error(e),
     complete: () => end()
  })
  return disposer
})
```
does alike thing in rx adapter:
https://github.com/cyclejs/cyclejs/blob/master/rx-adapter/src/index.ts#L15
does such problem exist with rx?

**Comment by milankinen** [comment]
> In most, the disposer needs to be returned before the emissions begin. 

As far as I know, `@most/create` should do exactly that stuff behalf of you, e.g. see http://www.webpackbin.com/4JrCxmZFb

**Comment by TylorS** [comment]
@milankinen Is correct about `@most/create` the calls to the three functions inside the callback, in his example `next`, `error`, and `complete`, are not actually called until next tick.

**Comment by ntilwalli** [comment]
@TylorS Good to know.  @milankinen @whitecolor I updated the PR to use `@most/create` and removed the `setTimeout` usage.
